airbyte-integrations/connectors/destination-jdbc/src/test/java/io/airbyte/integrations/destination/jdbc/copy/s3/LegacyS3StreamCopierTest.java

```diff
@@ -0,0 +1,156 @@
+        sqlOperations
+    ) {
+      @Override
+      public void copyS3CsvFileIntoTable(
+          final JdbcDatabase database,
+          final String s3FileLocation,
+          final String schema,
+          final String tableName,
+          final S3DestinationConfig s3Config) {
+        throw new UnsupportedOperationException("not implemented");
+      }
+    };
+  }
+
+  @AfterEach
+  public void teardown() {
+    streamTransferManagerMockedConstruction.close();
+  }
+
+  @Test
+  public void createSequentialStagingFiles_when_multipleFilesRequested() {
+    // When we call prepareStagingFile() the first time, it should create exactly one upload manager
+    final String firstFile = copier.prepareStagingFile();
+    assertEquals("fake-staging-folder/fake-schema/fake-stream_00000", firstFile);
+    final List<StreamTransferManager> firstManagers = streamTransferManagerMockedConstruction.constructed();
+    final StreamTransferManager firstManager = firstManagers.get(0);
+    verify(firstManager.numUploadThreads(anyInt()).queueCapacity(anyInt())).partSize(PART_SIZE);
+    assertEquals(1, firstManagers.size(), "There were actually " + firstManagers.size() + " upload managers");
+
+    // Each file will contain multiple parts, so the first MAX_PARTS_PER_FILE will all go into the same file (i.e. we should not start more uploads)
+    // We've already called prepareStagingFile() once, so only go to MAX_PARTS_PER_FILE - 1
+    for (var i = 0; i < LegacyS3StreamCopier.MAX_PARTS_PER_FILE - 1; i++) {
+      final String existingFile = copier.prepareStagingFile();
+      assertEquals("fake-staging-folder/fake-schema/fake-stream_00000", existingFile, "preparing file number " + i);
+      final int streamManagerCount = streamTransferManagerMockedConstruction.constructed().size();
+      assertEquals(1, streamManagerCount, "There were actually " + streamManagerCount + " upload managers");
+    }
+
+    // Now that we've hit the MAX_PARTS_PER_FILE, we should start a new upload
+    final String secondFile = copier.prepareStagingFile();
+    assertEquals("fake-staging-folder/fake-schema/fake-stream_00001", secondFile);
+    final List<StreamTransferManager> secondManagers = streamTransferManagerMockedConstruction.constructed();
+    final StreamTransferManager secondManager = secondManagers.get(1);
+    verify(secondManager.numUploadThreads(anyInt()).queueCapacity(anyInt())).partSize(PART_SIZE);
+    assertEquals(2, secondManagers.size(), "There were actually " + secondManagers.size() + " upload managers");
+  }
+
+  @Test
+  public void closesS3Upload_when_stagingUploaderClosedSuccessfully() throws Exception {
+    copier.prepareStagingFile();
+
+    copier.closeStagingUploader(false);
+
+    final List<StreamTransferManager> managers = streamTransferManagerMockedConstruction.constructed();
+    final StreamTransferManager manager = managers.get(0);
+    verify(manager).numUploadThreads(10);
+    verify(manager).complete();
```

Comment:
@sherifnada this is the call that I mentioned - `numUploadThreads` is verified fine, but for some reason mockito refuses to acknowledge that `complete` is being called.

(`closesS3Upload_when_stagingUploaderClosedFailingly` is probably the structure I'll go with, if I don't figure this out shortly)

Reply:
for posterity: RETURNS_DEEP_STUBS is not friendly for mocking builder objects. RETURNS_SELF is slightly better. but in this specific case, because we're mocking a constructor, ended up passing in a mock initializer.